python-threatexchange/threatexchange/stopncii/api.py

```diff
@@ -0,0 +1,251 @@
+    caseNumbers: t.Dict[str, StopNCIICaseStatus]  # Individual cases that correspond
+    # Feedback on hashes by CSPs
+    CSPFeedbacks: t.List[StopNCIICSPFeedback] = field(default_factory=list)
+
+
+@dataclass
+class FetchHashesResponse:
+    """
+    Wrapper around the FetchHashes call response.
+
+    Advantages of using a dataclass is the typing!
+    """
+
+    count: int  # How many records are there?
+    nextPageToken: str  # Cursor for paginating, not valid over long periods of time
+    nextSetTimestamp: int  # The best timestamp to use to store as a checkpoint
+    hasMoreRecords: bool  # If the cursor is fully played out
+    hashRecords: t.List[StopNCIIHashRecord]  # The records
+
+
+class StopNCIIAPI:
+    """
+    A wrapper around the StopNCII.org hash exchange API.
+
+    Hashes are submitted by individual people to the portal at StopNCII.org,
+    and Content Service Providers (CSPs), such as social networks, can provide
+    feedback on the hashes on whether they are able to verify that the content
+    corresponds to NCII content.
+    """
+
+    BASE_URL: t.ClassVar[str] = "https://api.stopncii.org/"
+    API_VERSION: t.ClassVar[str] = "v1"
+
+    DEFAULT_START_TIMESTAMP: t.ClassVar[int] = 10
+    PAGE_SIZE: t.ClassVar[int] = 800
+
+    def __init__(self, function_key: str, subscription_key: str) -> None:
+        self._auth_headers = {
+            "x-functions-key": function_key,
+            "Ocp-Apim-Subscription-Key": subscription_key,
+        }
+
+    def _get_session(self):
+        """
+        Custom requests sesson
+
+        Ideally, should be used within a context manager:
+        ```
+        with self._get_session() as session:
+            session.get()...
+        ```
+
+        If using without a context manager, ensure you end up calling close() on
+        the returned value.
+        """
+        session = requests.Session()
+        session.headers.update(self._auth_headers)
+        session.mount(
+            self.BASE_URL,
+            adapter=TimeoutHTTPAdapter(
+                timeout=60,
+                max_retries=Retry(
+                    total=4,
+                    status_forcelist=[429, 500, 502, 503, 504],
+                    allowed_methods=["HEAD", "GET", "OPTIONS"],
+                    backoff_factor=0.2,  # ~1.5 seconds of retries
+                ),
+            ),
+        )
+        return session
+
+    def _get(self, endpoint: str, **json) -> t.Any:
+        """
+        Perform an HTTP GET request, and return the JSON response payload.
+        Same timeouts and retry strategy as `_get_session` above.
+        """
+
+        url = "/".join((self.BASE_URL, self.API_VERSION, endpoint))
+        with self._get_session() as session:
+            response = session.get(url, json=json)
+            response.raise_for_status()
+            return response.json()
+
+    def _post(self, endpoint: str, *, json=None, params=None) -> t.Any:
```

Comment:
same: `_post_json_or_fail`?

Reply:
I prefer the shorter names, and think they would not be surprising to a future person. Should add docstring though